### PR TITLE
Send plug telemetry events

### LINKED ISSUE #1: Test Plug.AMQP module
`Plug.AMQP` module is not tested at all. At least `handle/4` should be tested.

### LINKED ISSUE #2: Support (plug) expected telemetry events
We are not sending the telemetry event that plug expect. More info [here](https://github.com/elixir-plug/plug/blob/master/lib/plug/conn/adapter.ex#L5).

Related [TODO](https://github.com/kantox/plug_amqp/blob/master/lib/plug/amqp/consumer_producer.ex#L34).

### PR DESCRIPTION
**Pull request by saverio-kantox**
Fixes #2 
Fixes #1

Also fixes sending a resp when the plug doesn't explicitly call `Plug.Conn.send_resp`.

### CODE REVIEW

File: lib/plug/amqp.ex [hunk 1]
```diff
@@ -98,4 +135,29 @@ defmodule Plug.AMQP do
...
+  defp exit_on_error(:throw, value, stack, call) do
+    exit({{{:nocatch, value}, stack}, call})
+  end
+
+  defp exit_on_error(:exit, value, _stack, call) do
+    exit({value, call})
+  end
+
+  defp maybe_send_resp(%Plug.Conn{state: :set} = conn), do: Plug.Conn.send_resp(conn)
+  defp maybe_send_resp(%Plug.Conn{} = conn), do: conn
```
[the remark this answers is not in the document]
Comment (saverio-kantox) [reply lib/plug/amqp.ex:162]:
```suggestion
  defp maybe_send_resp(conn = %Plug.Conn{}), do: conn
```

File: lib/plug/amqp.ex [hunk 2]
```diff
@@ -98,4 +135,29 @@ defmodule Plug.AMQP do
...
+
+  defp exit_on_error(:throw, value, stack, call) do
+    exit({{{:nocatch, value}, stack}, call})
+  end
+
+  defp exit_on_error(:exit, value, _stack, call) do
+    exit({value, call})
+  end
+
+  defp maybe_send_resp(%Plug.Conn{state: :set} = conn), do: Plug.Conn.send_resp(conn)
```
[the remark this answers is not in the document]
Comment (saverio-kantox) [reply lib/plug/amqp.ex:161]:
```suggestion
  defp maybe_send_resp(conn = %Plug.Conn{state: :set}), do: Plug.Conn.send_resp(conn)
```

File: test/plug/amqp/handler_test.exs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+defmodule Plug.AMQPTest do
+  use ExUnit.Case, async: true
+
+  defmodule MyPlug do
+    def call(%Plug.Conn{path_info: ["foo", "bar"]} = conn, _) do
+      Plug.Conn.resp(conn, 200, "pong")
+    end
+
+    def call(%Plug.Conn{path_info: ["force", "bar"]} = conn, _) do
```
[the remark this answers is not in the document]
Comment (saverio-kantox) [reply test/plug/amqp/handler_test.exs:9]:
```suggestion
    def call(conn = %Plug.Conn{path_info: ["force", "bar"]}, _) do
```

File: test/plug/amqp/handler_test.exs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+defmodule Plug.AMQPTest do
+  use ExUnit.Case, async: true
+
+  defmodule MyPlug do
+    def call(%Plug.Conn{path_info: ["foo", "bar"]} = conn, _) do
```
[the remark this answers is not in the document]
Comment (saverio-kantox) [reply test/plug/amqp/handler_test.exs:5]:
```suggestion
    def call(conn = %Plug.Conn{path_info: ["foo", "bar"]}, _) do
```

File: test/plug/amqp/telemetry_test.exs [hunk 5]
```diff
@@ -0,0 +1,95 @@
+defmodule Plug.AMQP.TelemetryTest do
+  use ExUnit.Case, async: true
+
+  def attach_telemetry() do
```
[the remark this answers is not in the document]
Comment (saverio-kantox) [reply test/plug/amqp/telemetry_test.exs:4]:
```suggestion
  def attach_telemetry do
```

File: test/plug/amqp/handler_test.exs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  test "handle routed message, force send response" do
+    assert :ok =
+             Plug.AMQP.handle(self(), "ping", [{"amqp-routing-key", "force.bar"}], plug: MyPlug)
+
+    assert_receive {:send_resp, _, "pong", _}
+  end
+
+  test "handle message that would raise but still sends a resp" do
+    Plug.AMQP
+    |> Task.start(:handle, [
```
Comment (alvivi) [inline test/plug/amqp/handler_test.exs:40]: A single pipeline is a lonely pipeline:

```
Task.start(Plug.AMQP, :handle, [
```

File: test/plug/amqp/telemetry_test.exs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                    }}
+
+    refute_received {:event, [:plug_adapter, :call, :exception], _, _}
+  end
+
+  test "emits telemetry events for start/exception" do
+    attach_telemetry()
+
+    Plug.AMQP
+    |> Task.start(:handle, [
```
Comment (alvivi) [inline test/plug/amqp/telemetry_test.exs:71]: Another sad pipe.

File: test/plug/amqp/telemetry_test.exs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      Plug.Conn.send_resp(conn, 200, "STOP")
+      raise "oops"
+    end
+  end
+
+  test "emits telemetry events for start/stop" do
+    attach_telemetry()
+
+    Plug.AMQP
+    |> Task.async(:handle, [
```
Comment (alvivi) [inline test/plug/amqp/telemetry_test.exs:41]: A pipe with no friends.

File: test/plug/amqp/telemetry_test.exs [hunk 9]
```diff
@@ -0,0 +1,95 @@
+defmodule Plug.AMQP.TelemetryTest do
+  use ExUnit.Case, async: true
+
+  def attach_telemetry do
+    unique_name = :"PID#{System.unique_integer()}"
```
Comment (alvivi) [inline test/plug/amqp/telemetry_test.exs:5]: Why do you need to care about the name/registration. Why not just use a closure:

```
handler = self()
```
 
and then:

```
send(handler, {:event, event, measurements, metadata})
```

File: test/plug/amqp/telemetry_test.exs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    Process.register(self(), unique_name)
+
+    for suffix <- [:start, :stop, :exception] do
+      :telemetry.attach(
```
Comment (alvivi) [inline test/plug/amqp/telemetry_test.exs:9]: You can use `telemetry:attach_many/4` for this.

File: lib/plug/amqp/consumer_producer.ex [hunk 11]
```diff
@@ -394,7 +391,7 @@ defmodule Plug.AMQP.ConsumerProducer do
   #
 
   @impl true
-  def handle_cast(
+  def handle_info(
```
Comment (alvivi) [inline lib/plug/amqp/consumer_producer.ex:394]: If we change this to `handle_info` we should move this to it proper place (next to the other `handle_info`) and also move the above comment and probably remove the `@impl true`. Also from:

```
  #
  # Server - Handle Cast
  #
```

To something like:

```
  #
  # Server - Handle Info - Send Resp (or client calls, or whatever)
  #
```

File: test/plug/amqp/telemetry_test.exs (hunk 9 above)
Comment (saverio-kantox) [reply test/plug/amqp/telemetry_test.exs:5]: It does not work, the telemetry events leak across tests.

Comment (saverio-kantox) [reply test/plug/amqp/telemetry_test.exs:5]: Actually my version is flaky too. Telemetry events from other test files are being captured by this handler.

**Pull request merged**